### PR TITLE
New block production algorithm

### PR DESCRIPTION
**Pull request by macgregordennis**
This new algorithm does the following:

- Get pending transactions
- Sort them in ascending order of invalidity count (new field) and descending order of transaction fee.
- Incrementally add transactions to a pending block until a target number of transactions per block is reached.
- With each additional transaction, check whether the pending block with the new tx is valid. If it is valid, proceed with the newly added transaction as part of the pending block. If it is invalid, increment the tx's local invalidity count and proceed with the last valid block.
- Return the last valid pending block

This algorithm is an improvement to block production since it applies all block validations to the combined set of transactions being considered for a new block rather than considering the validity of each transaction in isolation.

### CODE REVIEW

File: lib/credo_core_node/blockchain/block_producer.ex [hunk 1]
```diff
@@ -49,19 +52,54 @@ defmodule CredoCoreNode.Blockchain.BlockProducer do
...
-    batch = txs || Pool.get_batch_of_valid_pending_transactions()
-
-    if length(batch) > 0 do
-      batch
-      |> BlockReward.add_block_reward_tx()
-      |> Pool.generate_pending_block()
+    pending_block =
+      (txs || Pool.list_pending_transactions())
+      |> Enum.sort(
+        &((&1.invalidity_count || 0) <= (&1.invalidity_count || 0) &&
```
Comment (malroc) [inline lib/credo_core_node/blockchain/block_producer.ex]: @macgregordennis this won't work.
First, there is a typo: `&((&1.invalidity_count || 0) <= (&1.invalidity_count || 0)` (please notice the second `&1`).
Second, the sorting function is invalid:
```elixir
iex(20)> lst = (1..20) |> Enum.map(fn _ -> %{invalidity_count: Enum.random(1..10), fee: Enum.random(1..10)} end)                                                        [
  %{fee: 7, invalidity_count: 3},
  %{fee: 6, invalidity_count: 4},
  %{fee: 1, invalidity_count: 4},
  %{fee: 3, invalidity_count: 10},
  %{fee: 6, invalidity_count: 9},
  %{fee: 9, invalidity_count: 5},
  %{fee: 4, invalidity_count: 5},
  %{fee: 1, invalidity_count: 4},
  %{fee: 3, invalidity_count: 2},
  %{fee: 2, invalidity_count: 7},
  %{fee: 9, invalidity_count: 10},
  %{fee: 7, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 7, invalidity_count: 10},
  %{fee: 3, invalidity_count: 5},
  %{fee: 5, invalidity_count: 6},
  %{fee: 3, invalidity_count: 3},
  %{fee: 4, invalidity_count: 7},
  %{fee: 10, invalidity_count: 1}
]
iex(21)> lst |> Enum.sort(&((&1.invalidity_count) <= (&2.invalidity_count) && (&1.fee >= &2.fee)))
[
  %{fee: 10, invalidity_count: 1},
  %{fee: 7, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 9, invalidity_count: 5},
  %{fee: 9, invalidity_count: 10},
  %{fee: 7, invalidity_count: 3},
  %{fee: 7, invalidity_count: 10},
  %{fee: 3, invalidity_count: 2},
  %{fee: 3, invalidity_count: 3},
  %{fee: 6, invalidity_count: 4},
  %{fee: 5, invalidity_count: 6},
  %{fee: 4, invalidity_count: 5},
  %{fee: 4, invalidity_count: 7},
  %{fee: 3, invalidity_count: 5},
  %{fee: 2, invalidity_count: 7},
  %{fee: 1, invalidity_count: 4},
  %{fee: 6, invalidity_count: 9},
  %{fee: 3, invalidity_count: 10},
  %{fee: 1, invalidity_count: 4}
]
```
Notice that the resulting list is sorted semi-randomly.
What you need is either:
```elixir
iex(22)> lst |> Enum.sort(&((&1.invalidity_count) < (&2.invalidity_count) || ((&1.invalidity_count) == (&2.invalidity_count) && (&1.fee >= &2.fee))))
[
  %{fee: 10, invalidity_count: 1},
  %{fee: 3, invalidity_count: 2},
  %{fee: 7, invalidity_count: 3},
  %{fee: 7, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 3, invalidity_count: 3},
  %{fee: 6, invalidity_count: 4},
  %{fee: 1, invalidity_count: 4},
  %{fee: 1, invalidity_count: 4},
  %{fee: 9, invalidity_count: 5},
  %{fee: 4, invalidity_count: 5},
  %{fee: 3, invalidity_count: 5},
  %{fee: 5, invalidity_count: 6},
  %{fee: 4, invalidity_count: 7},
  %{fee: 2, invalidity_count: 7},
  %{fee: 6, invalidity_count: 9},
  %{fee: 9, invalidity_count: 10},
  %{fee: 7, invalidity_count: 10},
  %{fee: 3, invalidity_count: 10}
]
```
to sort by `invalidity_count` first, or:
```elixir
iex(23)> lst |> Enum.sort(&(((&1.invalidity_count) <= (&2.invalidity_count) && (&1.fee == &2.fee)) || (&1.fee > &2.fee)))
[
  %{fee: 10, invalidity_count: 1},
  %{fee: 9, invalidity_count: 5},
  %{fee: 9, invalidity_count: 10},
  %{fee: 7, invalidity_count: 3},
  %{fee: 7, invalidity_count: 3},
  %{fee: 7, invalidity_count: 10},
  %{fee: 6, invalidity_count: 4},
  %{fee: 6, invalidity_count: 9},
  %{fee: 5, invalidity_count: 3},
  %{fee: 5, invalidity_count: 3},
  %{fee: 5, invalidity_count: 6},
  %{fee: 4, invalidity_count: 5},
  %{fee: 4, invalidity_count: 7},
  %{fee: 3, invalidity_count: 2},
  %{fee: 3, invalidity_count: 3},
  %{fee: 3, invalidity_count: 5},
  %{fee: 3, invalidity_count: 10},
  %{fee: 2, invalidity_count: 7},
  %{fee: 1, invalidity_count: 4},
  %{fee: 1, invalidity_count: 4}
]
```
to sort by `fee` first

Comment (macgregordennis) [reply lib/credo_core_node/blockchain/block_producer.ex]: Good catches. I'll fix those issues shortly.

Comment (macgregordennis) [reply lib/credo_core_node/blockchain/block_producer.ex]: Fee still needs decimal comparison though

Comment (malroc) [reply lib/credo_core_node/blockchain/block_producer.ex]: @macgregordennis right, I used integer values/comparison operator in the example code for simplicity only. I also omitted ` || 0` part for the same purpose but you likely should add it back.

**Pull request merged**